WIP: Isotopic data

Gregstrq opened the pull request:
This PR adresses issue #37.
It adds new structs called `Isotope` and `Isotopes` which are analogues of `Element` and `Elements`. Also it defines the object `isotopes::Isotopes` which stores all the isotopic data. The indexing into `Isotopes` is based around the combination of the name or id of the element with the mass number. For example, to access tritium one can use either of the following options:
```
isotopes[:H3]
isotopes[:H, 3]
isotopes[1, 3]
isotopes["hydrogen", 3]
isotopes["hydrogen3"]
```

It is also important to be able to access isotopes corresponding to the element, and vice versa. As such, `isotopes[elements[:H]]` returns an array, consisting of H1, deuterium and tritium. Analogously, `elements[isotopes[:H3]]` returns the hydrogen.

**Important Disclaimer**
I have generated the isotopic data as described in [https://github.com/Gregstrq/Isotope-data](https://github.com/Gregstrq/Isotope-data). The scripts I used are stored there as well.
Currently, there is data only for 323 isotopes. The problem is, I pulled the data from the different sources, and the weakest link is the [easyspin database](https://easyspin.org/documentation/isotopetable.html), which I used to get the data for spins, g-factors and quadrupole moments. I can get the data for a much larger set of isotopes, but for that I need to parse the table from a pdf-file. I don't know how to do that. (I list this pdf sources on [repo page](https://github.com/Gregstrq/Isotope-data).)

I have checked that it workes locally in my REPL. If the maintainers like this PR, I will work on the tests.

TimSlendebroek commented:
This feature would be very useful for me, I hope this could be merged. Thanks for working on this!

carstenbauer commented on src/isotopes_data.jl:
```diff
@@ -0,0 +1,4847 @@
+_isotopes_data = [
```
Should be `const`?

carstenbauer replied:
@Gregstrq ☝️

carstenbauer commented on src/isotopes_data.jl:
```diff
@@ -0,0 +1,4847 @@
+_isotopes_data = [
+    Isotope(
+        "hydrogen",
+        "H",
+        1,
+        1,
+        99.9885,
+        1.007825031898*u,
+        1.4e-11*u,
+        1//2,
+        5.58569468,
+        0.0*b,
+        false,
+        Inf*s,
+        NaN*s
+    ),
```
For `Element` we define a constructor taking keyword arguments (see [here](https://github.com/JuliaPhysics/PeriodicTable.jl/blob/master/src/PeriodicTable.jl#L45-L70)). I think we should do the same for `Isotope` because 1) this makes it clear what the numbers in the definitions mean and makes them easier to parse for humans and 2) to be consistent.

Gregstrq replied:
Do we really need the humans to read it? This code is generated automatically from a DataFrame anyway.

But, I guess, I don't really care. Making the dictionary constructor does not require a lot of effort.

carstenbauer reviewed (changes requested):
I took a quick look at this PR and left some concrete comments.

On a more general note, I think that it would be nice to have the "sources issue" figured out. I would imagine that parsing a PDF, while annoying, should be doable with something like https://github.com/sambitdash/PDFIO.jl.

Moreover, I was wondering whether from an API point of view the functionality to "convert" from elements to isotopes and backwards should (additionally?) be provided as functions. Personally, I think that `isotopes(some_element)` and `element(some_isotope)` are more Julian. But, of course, that's a matter of taste. What do you think?

Gregstrq commented:
> Moreover, I was wondering whether from an API point of view the functionality to "convert" from elements to isotopes and backwards should (additionally?) be provided as functions.

The interface of elements is based on indexing, so for the reason of consistency, I kept the interfacing between elements and isotopes as the indexing operation as well. So, we can provide the functions, but additionally to the indexing interface.

carstenbauer commented:
Yeah, maybe it's just me and indexing is fine here. @rahulkp220 What's your opinion on this?

rahulkp220 commented:
Hey, got a bit late here. I think I am okay with indexing :)
@carstenbauer please feel free to merge.

anandijain commented:
Can make Isotopes.jl that depends on PeriodicTable.jl? also why are there no tests and why CI didn't run?

carstenbauer commented:
@Gregstrq Can you please add some tests?

@anandijain Good question why CI didn't run... it [should](https://github.com/JuliaPhysics/PeriodicTable.jl/blob/master/.github/workflows/CI.yml#L4)?!

Eben60 commented:
> Currently, there is data only for 323 isotopes. The problem is, I pulled the data from the different sources, and the weakest link is the easyspin database

We are not going to have **complete** data for ALL isotopes, so shouldn't it be OK to have some values as `missing`?

Eben60 commented:
>  I need to parse the table from a pdf-file. I don't know how to do that.

I have converted one of the tables to unicode tab-separated text. The workflow: 

Acrobat Pro: 

- export to MS Word (worked astoundingly well)

MS Word: 

- remove everything outside the tables
- replace mu and beta characters (represented as m and b in Symbol font upon conversion) to unicode Greek symbols in Arial font
- copy all to clipboard

MS Excel
- paste
- export as unicode

I hope there were no other non-Latin characters in the texts, and that I didn't otherwise introduced any errors. The files are in https://github.com/Eben60/isotopic_data , I'm going to convert another PDF file too, but not right now. I'd leave the further processing to CSV.jl experts ;)

Gregstrq commented:
@anandijain I can make "Isotopes.jl" a separate package. But then, the question is, is it really more convenient for others?

I actually contacted the guys from the committee that created the pdf files I mentioned. They sent me csv files with recommended values of magnetic dipole and electric quadrupole moments. I am currently reworking my PR to accommodate new data.

Gregstrq commented:
By the way, can anyone explain to me why `NaN`-s are used instead of `missing`-s in the elements data?

lmmentel commented:
Hi All, I'm the author of [mendeleev](https://github.com/lmmentel/mendeleev). I wasn't aware of this effort until @Eben60 created a PR linked above. This looks really great, kudos. If you are interested I already have some isotope data parsed from [CIAAW](http://ciaaw.org/atomic-masses.htm.) into a sqlite db. 

I wonder if there's some synergies here we could explore in order to not duplicate the efforts and share data?

Gregstrq commented:
@Eben60, @Immentel I have updated https://github.com/Gregstrq/Isotope-data repo. The dataframe with isotopic data is called "isotopes_data.jls". To load up the dataframe, just do
```
using Serialization, FileIO, DataFrames, Unitful
isotopes = deserialize("isotopes_data.jls")
```
I did not save the data as a db because I wanted to keep the units of the physical data.

lmmentel commented:
> I did not save the data as a db because I wanted to keep the units of the physical data.

Got it. I am currently keeping track of the units only in the docs layer, i.e. docstrings and curated documentation but it would probably be better to have that information in the db as well, I'll create an issue for it over at [mendeleev](https://github.com/lmmentel/mendeleev).

As mentioned in [this comment](https://github.com/Gregstrq/Isotope-data/pull/1#issuecomment-1257297094) I had some issues opening the isotope data but went ahead and [parsed it myself](https://github.com/lmmentel/mendeleev/pull/74) and it should soon make it the the db and into the next release of [mendeleev](https://github.com/lmmentel/mendeleev).

Gregstrq commented:
I guess we need to close this PR.
@Eben60, @Immmentel, I have split this PR into a separate package: [IsotopeTable.jl](https://github.com/Gregstrq/IsotopeTable.jl). Additionally to that, I have created a package [IsotopeTableDF.jl](https://github.com/Gregstrq/IsotopeTableDF.jl), which simply exports the isotope data as a `DataFrame`. (I have split isotopes DataFrame into a separate package, because its loading from .jld2 file requires extra packages.)

The registration of these packages should be finished tomorrow. (3 day waiting period.)

lmmentel commented:
Thanks for the update @Gregstrq. I have parsed the isotopic data into a SQLite db that is a part of [mendeleev](https://github.com/lmmentel/mendeleev) in case anyone is interested in a language independent source of this data.
